Add MongoDB 3.2 container image.

omron93 opened the pull request:
The latest stable MongoDB release is 3.2.

This PR adds a container image for this release. 3.2 implementation/directory is really similar to 2.6 (only necessary installation changes, db.addUser to db.createUser substitution in test/run).

**Image from this PR is not buildable!**
The rh-mongodb32 SCL is not build in CentOS yet. I am creating this PR to discuss some thing to be able to make a quicker merge after builds in cbs will be ready.

One big change in MongoDB 3.2 is a new default storage engine - WiredTiger. This storage engine bring new feature and better performance to MongoDB (https://www.mongodb.com/blog/post/whats-new-mongodb-30-part-3-performance-efficiency-gains-new-storage-architecture), but it is x86_64 only!
For WiredTiger MONGODB_NOPREALLOC and MONGODB_SMALLFILES has no effect. WT behaves similar by default.

@mfojtik Is it a problem that WT is 64bit only? Could I remove MONGODB_NOPREALLOC and MONGODB_SMALLFILES options from README, usage() and config file?

omron93 commented:
@mfojtik @bparees ping

mfojtik commented:
@omron93 I think our strategy is to always keep rhel7 + centos7 image together so I would wait merging this till the centos7 SCL is ready.

Is there any delta in scripts between this version and the existing versions? Also have you tested replication?

[test]

omron93 commented:
> I think our strategy is to always keep rhel7 + centos7 image together so I would wait merging this till the centos7 SCL is ready.

I understand. Now I only want to know OpenShift POV on storage engine changes and if it is possible to remove config options... [see my comment](https://github.com/openshift/mongodb/pull/127#issue-134040712)

> Is there any delta in scripts between this version and the existing versions?

[rh-mongodb-32-patch.txt](https://github.com/openshift/mongodb/files/136268/rh-mongodb-32-patch.txt)

> Also have you tested replication?

No. I thought it is supported only in mongodb-24.
Replication scripts are same as for mongodb-26.

omron93 commented:
Changed installation in CentOS Dockerfile. It is now using centos-sclo-rh-testing repository... (see https://github.com/openshift/postgresql/pull/99#issuecomment-185416629)

However it is not working now. It depends on this bug https://bugs.centos.org/view.php?id=10302

bparees commented:
@omron93 i think 64-bit only is ok, and you can remove the ENV variables that have will no longer have any effect.

@mfojtik let's not hold this up on replication.  Once @rhcarvalho fixes the replication for 2.4 and 2.6, and @php-coder writes replication tests, we can come back and test this with that.

bparees commented:
[test]

mfojtik commented:
LGTM (when it builds ;-)

rhcarvalho commented:
I feel really bad about our approach of copy-pasting files when adding support for new versions, for me it brings unnecessary maintenance burden.

@bparees if we're merging with the second commit in, we need an issue open to track moving away from a testing repository to an official release repository.

rhcarvalho commented:
@omron93 why don't we need the `v8314` package for 3.2, is it also because of changes in MongoDB?

rhcarvalho commented:
The diff between `2.6` and `3.2` looks reasonable, only three files differ:
- `test/run`
- `Dockerfile`
- `Dockerfile.rhel`

The changes in `test/run` are `removeUser -> dropUser` and `addUser -> createUser`. Perhaps those should be changed in the replication scripts as well, to save us time figuring out why the scripts don't work when we will be able to support replication for 3.2.

bparees commented:
> I feel really bad about our approach of copy-pasting files when adding support for new versions, for me it brings unnecessary maintenance burden.

@rhcarvalho i'm open to suggestions, but trying to make this stuff common and then finding parts of it are not common, is also a large maintenance burden.  We are also not likely to have that many versions of any given image, especially with older versions rolling EOL.

bparees commented:
@rhcarvalho one option i'd consider, if all the files are really identical, would be creating symlinks for now, and breaking them if we have to introduce deltas.

bparees commented:
> @bparees if we're merging with the second commit in, we need an issue open to track moving away from a testing repository to an official release repository.

yes.  @omron93 do you want to open that issue?

mfojtik commented:
@bparees I don't think symlinks with work with Docker build (just guessing)

mfojtik commented:
@bparees but `make build` can copy the files if needed.

omron93 commented:
> why don't we need the v8314 package for 3.2, is it also because of changes in MongoDB

MongoDB 3.2 switched from v8 to mozjs38 JS engine

> The changes in test/run are removeUser -> dropUser and addUser -> createUser. Perhaps those should be changed in the replication scripts as well, to save us time figuring out why the scripts don't work when we will be able to support replication for 3.2.

This deprecated functions are replaced even in 2.6 scripts -> so there is no change in it.

> yes. @omron93 do you want to open that issue?

New issue - https://github.com/openshift/mongodb/issues/129

bparees commented:
@omron93 remove the references to the unused env variables, squash, and i'll merge (if it passes tests)

omron93 commented:
> remove the references to the unused env variables, squash, and i'll merge (if it passes tests)

DONE

New diff from 2.6 - [rh-mongodb-32-patch.txt](https://github.com/openshift/mongodb/files/137817/rh-mongodb-32-patch.txt)

However this PR still have to wait for CentOS builds availability :-(

bparees commented:
@omron93 you don't want it published using the content of the test repo?

omron93 commented:
> you don't want it published using the content of the test repo?

Yes, I want to. However it was not ready even in the testing repo.
Now hhorak is building rh-mongodb32 in cbs, so hopefully it will be available in the testing repo soon.

bparees commented:
@omron93 ok, re-ping me when the centos and rhel repos are both ready and kick off a test for this.

elyscape commented on 3.2/README.md:
```diff
@@ -0,0 +1,77 @@
+Usage
+---------------------------------
+
+For this, we will assume that you are using the `centos/mongodb-26-centos7` image.
```
This should probably be changed to `centos/mongodb-32-centos7`.

rhcarvalho replied:
good catch... we shouldn't really be copy-pasting files to add new versions :cry:

bparees replied:
@omron93 yes this still needs to be updated.

omron93 replied:
I missed this. Sry.

zipzit commented:
Hmmm.  I'm trying to link from OpenShift server --> mlab.com mongodb.  Keep getting auth errors.  [Reference. ](http://stackoverflow.com/questions/35935531/failed-database-connection-between-openshift-server-and-mlab-com)  Folks at mlabs claim:

> You'll need to upgrade your mongo shell version to 3.0+ in order to connect and authenticate to an mLab Sandbox database server. It looks like version 2.4.9 is being used. Can you please upgrade and try again?

Will fixing this issue have the effect of upgrading the mongo tools accessible on OpenShift?  Many thanks.

rhcarvalho commented:
> Will fixing this issue have the effect of upgrading the mongo tools accessible on OpenShift?

@zipzit this repository is for the MongoDB image for OpenShift v3. The problem you are facing is with OpenShift v2, so this PR is unrelated.

Please allow some time for somebody to answer your question on StackOverflow, or try other support channels.

rhcarvalho commented:
@omron93 any updates on the MongoDB 3.2 package availability?

Commit:
Add MongoDB 3.2 container image.
Use centos sclo testing repository.
Remove mmapv1 NOPREALLOC and SMALLFILES options.

Fixed README

Rebased to upstream/master.

Use NOPRALOC and SMALLFILES only for mounted volumes from older MongoDB versions.

Fixed default testing image.

Merged changes from upstream.

Using RHSCL GA packages.

omron93 commented:
@mfojtik @rhcarvalho Packages seems to be available. Please test and merge.

Differences from 2.6 are only very small. See [26-to-32-diff.txt](https://github.com/openshift/mongodb/files/213368/26-to-32-diff.txt)

bparees commented:
@omron93 I think you need to rebase, i'm seeing a lot more differences in the /root/usr/share/container-scripts/mongodb directory, where those scripts have been significantly improved in the 2.4 and 2.6 trees.

bparees commented:
also mongodb.conf.template now includes 2 new config settings (prealloc and smallfiles)

omron93 commented:
@bparees Rebased.

Diff from 2.6 version to 3.2 version - [26-to-32-diff.txt](https://github.com/openshift/mongodb/files/214990/26-to-32-diff.txt)

Config file fixed.

bparees commented:
@omron93 do the prealloc and smallfiles config options not apply to 3.2?

bparees commented:
[test]

bparees commented:
```
The command '/bin/sh -c yum install -y yum-utils &&     yum-config-manager --enable rhel-server-rhscl-7-rpms &&     yum-config-manager --enable rhel-7-server-optional-rpms &&     INSTALL_PKGS="bind-utils gettext iproute rsync tar rh-mongodb32-mongodb rh-mongodb32" &&     yum install -y --setopt=tsflags=nodocs $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     yum clean all &&     mkdir -p /var/lib/mongodb/data && chown -R mongodb.0 /var/lib/mongodb/ /var/opt/rh/rh-mongodb32/lib/mongodb &&     chmod g+w -R /var/opt/rh/rh-mongodb32/lib/mongodb &&     chmod -R g+rwx /var/lib/mongodb' returned a non-zero code: 2
```

guessing we are still missing the mongodb32 packages for rhel (maybe missing in rhui, @tdawson ? )

omron93 commented:
> do the prealloc and smallfiles config options not apply to 3.2?

Yes.
MongoDB ships two storage engines:
- mmapv1 - used in previous versions as default. It uses prealloc and smallfiles options.
- wiredTiger - default storage engine since 3.2. It brings performance improvements. It does not use prealloc and smallfiles options.

> guessing we are still missing the mongodb32 packages for rhel

RHSCL is not released yet :-(

omron93 commented:
> do the prealloc and smallfiles config options not apply to 3.2?

@bparees I was wrong, they apply to 3.2 with mounted volume created by 2.6 container. MongoDB will detect that the database was created by mmapv1 and use it. (To note: wiredTiger and mmapv1 are incompatible - it is necessary to run `mongodump` and `mongorestore` procedure)

Now these two options are marked only as compatible options for mounted volumes.

[26-to-32-diff.txt](https://github.com/openshift/mongodb/files/223703/26-to-32-diff.txt)

bparees commented:
@omron93 needs a squash, otherwise lgtm pending the availability of the 3.2 packages.

rhcarvalho commented:
@omron93 will there be an image / SCL release including MongoDB 3.0?

Looks like to upgrade from 2.6 to 3.2, one needs to first upgrade to 3.0:
https://docs.mongodb.org/manual/release-notes/3.2-upgrade/#upgrade-standalone-mongod-instance-to-mongodb-3-2

omron93 commented:
>  will there be an image / SCL release including MongoDB 3.0?
> 
> Looks like to upgrade from 2.6 to 3.2, one needs to first upgrade to 3.0:
> https://docs.mongodb.org/manual/release-notes/3.2-upgrade/#upgrade-standalone-mongod-instance-to-mongodb-3-2

In RHSCL there will be 3.0 package exactly for this purpose. This collection does not include things not necessary for the upgrade and it won't be supported for normal use.

It there some support for upgrades in OpenShift images?

bparees commented:
@omron93 we haven't faced an upgrade scenario yet.  There are a number of ways a user might tackle it, but certainly one option would be to change the imagestreamtag that a particular deployment config was using, so the container would get re-deployed w/ the new image.  So in this case someone who had a running mongo2.6 container would change their tag to mongo:3.0, let it redeploy/migrate, then change the tag to mongo3.2.  (this assumes their data is stored on a persistent volume of course).

but to do that we'd need to provide an explicit 3.0 image as well.

omron93 commented:
> but to do that we'd need to provide an explicit 3.0 image as well.

@bparees I understand. It shouldn't be a problem. Scripts for rh-mongodb30upg should be the same as for rh-mongodb32. So minimal changes.

How to specify that mongodb-30-centos7 should be used for upgrade only. Probably bugs which does not affect upgrade procedure to 3.2 won't be fixed. How does OpenShift handles thinks like this? One line in README is enought?

omron93 commented:
Squashed.

bparees commented:
> How to specify that mongodb-30-centos7 should be used for upgrade only. Probably bugs which does not affect upgrade procedure to 3.2 won't be fixed. How does OpenShift handles thinks like this? One line in README is enought?

going to have to think about that, and maybe pull in @brenton from productization.  We haven't had this situation before.

brenton commented:
I agree this is a someone unique situation.  Does the upgrade from 2.6->3.0->3.2 require manual intervention or downtime?

If it was possible to ship an image that was only usable for upgrade we could release it.  When I say "only usable for upgrade" I'm meaning if you tried to use the 3.0 image it would report a warning or something of that nature saying that it's only supported as a migration path to 3.2.

omron93 commented:
No manul interventions required. There are few compatibility changes between these versions, but user could be able to fix them (if needed) with 2.6 and then upgrade. To note in case 2.4->2.6 upgrade upgrade situation was little bit more complicated :-)

The upgrade collection is called rh-mongodb30upg. How the upgrade image should be called? Normally centos/mongodb-30-centos7 or centos/mongodb-30upg-centos7?

hhorak commented:
I'd actually suggest `centos/mongodb-30-upg-centos7`, which is similar to what we already use for variants in thermostat and devtoolset images. Anyway, the RHEL-based images are now available as Beta, in `rhel-server-rhscl-7-beta-rpms` channel, so using this channel this PR can be finished.

rhcarvalho commented:
@omron93 before we merge this, please rebase and compare to 2.6, there are probably changes since  last update here.

omron93 commented:
@rhcarvalho @bparees @mfojtik Are you fine with `centos/mongodb-30-upg-centos7`? It looks good to me.

Merged changes from master. RHEL7 variant now use RHSCL Beta packages. Also squashed.
New diff - [26-to-32-diff.txt](https://github.com/openshift/mongodb/files/262957/26-to-32-diff.txt)

Please test.

bparees commented:
@omron93 yeah that name sounds fine. Unfortunately the PR tester is broken at the moment due to issues w/ the github PR tester, but as soon as we get that worked out i'll get back to this.

bparees commented:
PR tester is back online.
[test]

omron93 commented:
@hhorak @bparees Do you have any idea where could be a problem with installation of rh-mongodb32?

`14: curl#77 - "Problem with the SSL CA cert (path? access rights?)"` ?

hhorak commented:
When I try manually the problematic URL, I see browser reporting "invalid security certificate". Probably something wrong in OpenShift repos?

bparees commented:
guessing the rhel base image we use just doesn't have the right beta repos enabled?

the failure i see is just a yum install failure:
The command '/bin/sh -c yum install -y yum-utils &&     yum-config-manager --enable rhel-server-rhscl-7-beta-rpms &&     yum-config-manager --enable rhel-7-server-optional-rpms &&     INSTALL_PKGS="bind-utils gettext iproute rsync tar rh-mongodb32-mongodb rh-mongodb32" &&     yum install -y --setopt=tsflags=nodocs $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     yum clean all &&     mkdir -p /var/lib/mongodb/data && chown -R mongodb.0 /var/lib/mongodb/ /var/opt/rh/rh-mongodb32/lib/mongodb &&     chmod g+w -R /var/opt/rh/rh-mongodb32/lib/mongodb &&     chmod -R g+rwx /var/lib/mongodb' returned a non-zero code: 2

bparees commented:
and i should note that the ruby2.3 PR passed tests using the same repos, so are we sure the necessary packages are in the test repos?

bparees commented:
nm, not sure why the ruby2.3 PR passed at some point, but it's currently not building successfully either.

hhorak commented:
@bparees I still don't think we can handle this from withing images. From what I know about building on RHEL, it depends on how the host is configured and subscribed. You might want to check how the host is configured and what is inside `/usr/share/rhel/secrets/` on host, because that is used inside of images, IIUIC. The images only enable repositories, IMO.

bparees commented:
> @bparees I still don't think we can handle this from withing images. From what I know about building on RHEL, it depends on how the host is configured and subscribed. You might want to check how the host is configured and what is inside /usr/share/rhel/secrets/ on host, because that is used inside of images, IIUIC. The images only enable repositories, IMO.

we actually bake the right repos into our base rhel image that we build on top of and it does not currently include the scl beta repos.  i'm not totally sure i want it to, either, at risk of picking up things we don't want.  Is that an unfounded fear?

hhorak commented:
From my PoV, there is nothing to be afraid of. Any Dockerfile is responsible for enabling only repositories it needs -- so, if the base image included _disabled_ RHSCL Beta repos, upper layer can't use them unless explicitly enabling them. So, it seems safe to me to have such disabled RHSCL Beta repos baked even in the base image.

hhorak commented:
@omron93 we can use proper RHSCL repos now.

omron93 commented:
It is using RHCSL 2.2 packages now.
@bparees @mfojtik Please test?

Diff is same (only RHEL repo differ) 
[26-to-32-diff.txt](https://github.com/openshift/mongodb/files/302981/26-to-32-diff.txt)

bparees commented:
[test]

omron93 commented:
Sorry, I've forgotten to rebase to master.

See, #164
